posthog/hogql_queries/web_analytics/web_analytics_query_runner.py

```diff
@@ -143,3 +151,91 @@ def _refresh_frequency(self):
+        )
+
+    def _get_or_calculate_sample_ratio(self) -> SamplingRate:
+        if not self.query.sampling or not self.query.sampling.enabled:
+            return SamplingRate(numerator=1)
+        if self.query.sampling.forceSamplingRate:
+            return self.query.sampling.forceSamplingRate
+
+        cache_key = self._sample_rate_cache_key()
+        cached_response = get_safe_cache(cache_key)
+        if cached_response:
+            return SamplingRate(**cached_response)
+
+        # To get the sample rate, we need to count how many page view events there were over the time period.
+        # This would be quite slow if there were a lot of events, so use sampling to calculate this!
+
+        with self.timings.measure("event_count_query"):
+            event_count = parse_select(
+                """
+SELECT
+    count() as count
+FROM
+    events
+SAMPLE 1/1000
+WHERE
+    {where}
+                """,
+                timings=self.timings,
+                placeholders={
+                    "where": self.events_where_data_range(),
+                },
+            )
+
+        with self.timings.measure("event_count_query_execute"):
+            response = execute_hogql_query(
+                query_type="event_count_query",
+                query=event_count,
+                team=self.team,
+                timings=self.timings,
+            )
+
+        if not response.results or not response.results[0] or not response.results[0][0]:
+            return SamplingRate(numerator=1)
+
+        count = response.results[0][0] * 1000
+        fresh_sample_rate = _sample_rate_from_count(count)
+
+        cache.set(cache_key, fresh_sample_rate, settings.CACHED_RESULTS_TTL)
+
+        return fresh_sample_rate
+
+    @cached_property
+    def _sample_rate(self) -> SamplingRate:
+        return self._get_or_calculate_sample_ratio()
+
+    @cached_property
+    def _sample_ratio(self) -> ast.RatioExpr:
+        sample_rate = self._sample_rate
+        return ast.RatioExpr(
+            left=ast.Constant(value=sample_rate.numerator),
+            right=ast.Constant(value=sample_rate.denominator) if sample_rate.denominator else None,
+        )
+
+    def _unsample(self, n: Optional[int | float]):
+        if n is None:
+            return None
+
+        return (
+            n * self._sample_rate.denominator / self._sample_rate.numerator
+            if self._sample_rate.denominator
+            else n / self._sample_rate.numerator
+        )
+
+
+def _sample_rate_from_count(count: int) -> SamplingRate:
+    # Change the sample rate so that the query will sample about 100_000 to 1_000_000 events, but use defined steps of
+    # sample rate. These numbers are just a starting point, and we can tune as we get feedback.
+    sample_target = 10_000
+    sample_rate_steps = [1_000, 100, 10]
+
+    for step in sample_rate_steps:
+        if count / sample_target > step:
+            return SamplingRate(numerator=1, denominator=sample_target)
```

Comment:
Should the `denominator` always be `sample_target` here? Thought we'd return the `step` instead for the lower `count` values

Reply:
Good spot